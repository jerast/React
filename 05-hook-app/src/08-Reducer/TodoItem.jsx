export const TodoItem = ({ toDo, handleDeleteTodo, handleToggleTodo }) => {
	return (
		<div 
			className={`list-item ${ toDo.done ? ' done' : '' }`} 
			role="listitem"
		>
			<span 
				onClick={ () => handleToggleTodo(toDo.id) }
				role="span" 
			>
				{ toDo.description }
			</span>
			<button 
				onClick={ () => handleDeleteTodo(toDo.id) }
			>
				&times;
			</button>
		</div>
	);
};
