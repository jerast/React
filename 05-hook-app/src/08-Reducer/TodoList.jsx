import { TodoItem } from "./";

export const TodoList = ({ toDoList, handleDeleteTodo, handleToggleTodo }) => {
	return (
		<ul>
			{toDoList.map((todo) => (
				<TodoItem 
					key={ todo.id } 
					toDo={ todo } 
					handleDeleteTodo={ handleDeleteTodo } 
					handleToggleTodo={ handleToggleTodo }
				/>
			))}
		</ul>
	);
};
