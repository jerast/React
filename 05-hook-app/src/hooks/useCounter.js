import { useState } from "react";

export const useCounter = ( initialValue = 10 ) => {

	const [counter, setCounter] = useState( initialValue );

	const increment = ( steps = 1 ) => setCounter( counter => counter + steps );
	const decrement = ( steps = 1 ) => setCounter( counter => counter - steps );
	const reset = () => setCounter( initialValue );

	return {
		counter,
		increment,
		decrement,
		reset
	};
};