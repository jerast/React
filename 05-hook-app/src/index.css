@tailwind base;
@tailwind components;
@tailwind utilities;

@layer components {
	* {
		@apply box-border;
	}

	body {
		@apply bg-zinc-800;
	}
	#root {
		@apply w-full grid gap-3 justify-items-center p-5  text-white;
	}

	.title {
		@apply text-center text-4xl my-4;
	}
	.subtitle {
		@apply text-center text-2xl my-4;
	}

	button {
		@apply px-5 py-2 bg-zinc-500 text-white rounded-full disabled:opacity-50 disabled:cursor-default;
	}
	input {
		@apply bg-zinc-500 px-5 py-2 rounded-xl w-full;
	}

	blockquote {
		@apply w-full max-w-[500px];
	}
	blockquote p {
		@apply font-[cursive] text-6xl m-3 text-center;
	}
	blockquote span {
		@apply text-end w-full inline-block font-light font-mono italic text-sm;
	}
}
