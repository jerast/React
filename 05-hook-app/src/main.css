@tailwind base;
@tailwind components;
@tailwind utilities;

@layer components {
	* {
		@apply box-border select-none;
	}

	body {
		@apply bg-zinc-800;
	}
	#root {
		@apply w-full grid gap-3 justify-items-center p-5  text-white;
	}

	.title {
		@apply text-center text-4xl;
	}
	.subtitle {
		@apply text-center text-xl text-zinc-400;
	}

	button {
		@apply px-5 py-2 bg-zinc-700 text-white rounded-full disabled:opacity-50 disabled:cursor-default outline-none ease-in-out duration-75 hover:bg-zinc-800;
	}
	input {
		@apply bg-zinc-600 px-5 py-2 rounded-xl w-full;
	}
	form {
		@apply flex rounded-lg overflow-hidden shadow-md my-8
	}
	form > * {
		@apply rounded-none outline-none 
	}

	ul {
		@apply grid gap-2 w-full max-w-[500px]
	}
	.list-item {
		@apply w-full relative h-12 cursor-pointer ease-in-out duration-75
	}
	.list-item span {
		@apply w-full h-full rounded-full inline-flex items-center bg-zinc-500  hover:bg-zinc-600 ease-in-out duration-75 pl-5 shadow-md;
	}
	.list-item button {
		@apply text-2xl absolute right-0 px-5
	}
	.list-item.done {
		@apply  opacity-30;
	}
	.list-item.done span {
		@apply bg-zinc-600 line-through shadow-none;
	}

	blockquote {
		@apply w-full max-w-[500px];
	}
	blockquote p {
		@apply font-[cursive] text-6xl m-3 text-center;
	}
	blockquote span {
		@apply text-end w-full inline-block font-light font-mono italic text-sm;
	}
}
