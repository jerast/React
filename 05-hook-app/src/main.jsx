import React from 'react';
import ReactDOM from 'react-dom/client';
import { Padre } from './07-tarea-memo/Padre';
import './index.css';

// import { HooksApp } from './HooksApp'
// import { CounterApp } from './01-State/CounterApp'
// import { CounterWithCustomHook } from './01-State/CounterWithCustomHook';
// import { FormWithCustomHook } from './02-Effect/FormWithCustomHook';
// import { SimpleForm } from './02-Effect/SimpleForm';
// import { MultipleCustomHooks } from './03-Examples/MultipleCustomHooks';
// import { FocusScreen } from './04-Ref/FocusScreen';
// import { Layout } from './05-LayourEffect/Layout';
// import { Memorize } from './06-Memos/Memorize';
// import { MemoHook } from './06-Memos/MemoHook';
// import { CallbackHook } from './06-Memos/CallbackHook';

ReactDOM.createRoot(document.getElementById('root')).render(
	<>
		{/* <React.StrictMode> */}
			{/* <HooksApp /> */}
			{/* <CounterApp /> */}
			{/* <CounterWithCustomHook /> */}
			{/* <SimpleForm /> */}
			{/* <FormWithCustomHook /> */}
			{/* <MultipleCustomHooks /> */}
			{/* <FocusScreen /> */}
			{/* <Layout /> */}
			{/* <Memorize /> */}
			{/* <MemoHook /> */}
			{/* <CallbackHook /> */}
			<Padre />
		{/* </React.StrictMode> */}
	</>
);
