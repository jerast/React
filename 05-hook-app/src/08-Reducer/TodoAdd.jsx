import { useForm } from '../hooks';

export const TodoAdd = ({ handleCreateTodo }) => {

   const { description, onFormChange, onFormReset } = useForm({ description: '' });

   const onFormSubmit = (event) => {
      event.preventDefault();
      
      if ( !description ) return;

      const newTodo = {
         id: new Date().getTime(),
         description,
         done: false,
      }

      handleCreateTodo( newTodo );
      onFormReset();
   } 

	return (
		<form onSubmit={ onFormSubmit }>
			<input
				type="text"
				placeholder="What's to do?"
				name="description"
				value={ description }
				onChange={ onFormChange }
            required
			/>
			<button>+</button>
		</form>
	);
};
