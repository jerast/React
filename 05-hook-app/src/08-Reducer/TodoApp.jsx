import { useTodo } from "../hooks";
import { TodoAdd, TodoList } from "./";

export const TodoApp = () => {

   const { 
      todos,
      todosCount,
      pendingTodosCount,
      handleCreateTodo,
      handleDeleteTodo,
      handleToggleTodo 
   } = useTodo();

	return (
		<>
			<h1 className="title">ToDo App: { todosCount }</h1>
			<h1 className="subtitle">Pendings: { pendingTodosCount }</h1>

         <TodoAdd handleCreateTodo={ handleCreateTodo } />
         <TodoList 
            toDoList={ todos } 
            handleDeleteTodo={ handleDeleteTodo } 
            handleToggleTodo={ handleToggleTodo }
         />
		</>
	);
};
