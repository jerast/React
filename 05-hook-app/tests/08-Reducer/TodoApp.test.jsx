import { render, screen } from "@testing-library/react";
import { TodoApp } from "../../src/08-Reducer/TodoApp";
import { useTodo } from "../../src/hooks";

jest.mock('../../src/hooks/useTodo');

describe('Testing on <TodoApp /> component', () => {

   useTodo.mockReturnValue({
      todos: [
         { id: 1, description: 'Todo #1', done: false },
         { id: 2, description: 'Todo #2', done: true },
      ],
      todosCount: 2,
      pendingTodosCount: 1,
      handleCreateTodo: jest.fn(),
      handleDeleteTodo: jest.fn(),
      handleToggleTodo: jest.fn()
   })

   test('should display the component correctly', () => {
      const { container } = render( <TodoApp /> );

      expect( container ).toMatchSnapshot();
      expect( screen.getByText('Todo #1') ).toBeTruthy();
      expect( screen.getByText('Todo #2') ).toBeTruthy();
      expect( screen.getByRole('textbox') ).toBeTruthy();
   });

})