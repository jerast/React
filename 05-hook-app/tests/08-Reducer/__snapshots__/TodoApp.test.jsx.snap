// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing on <TodoApp /> component should display the component correctly 1`] = `
<div>
  <h1
    class="title"
  >
    ToDo App: 
    2
  </h1>
  <h1
    class="subtitle"
  >
    Pendings: 
    1
  </h1>
  <form>
    <input
      name="description"
      placeholder="What's to do?"
      required=""
      type="text"
      value=""
    />
    <button>
      +
    </button>
  </form>
  <ul>
    <div
      class="list-item "
      role="listitem"
    >
      <span
        role="span"
      >
        Todo #1
      </span>
      <button>
        ×
      </button>
    </div>
    <div
      class="list-item  done"
      role="listitem"
    >
      <span
        role="span"
      >
        Todo #2
      </span>
      <button>
        ×
      </button>
    </div>
  </ul>
</div>
`;
